internal/buf/pool.go

```diff
@@ -0,0 +1,205 @@
+// It is EXTREMELY important to always release memory allocated from the Pool. Failure to do so will
+// result in
+//
+// The pool uses N segments, with each segment tracking its high water mark usage.
+//
+// Allocation simply advances the high water mark within first segment that has capacity
+// and increments per-segment refcount.
+//
+// On Buf.Release() the refcount is decremented and when it hits zero, the entire segment becomes instantly
+// freed.
+//
+// As an extra optimization, when Buf.Release() is called in LIFO order, it will also lower the
+// high water mark making its memory available for immediate reuse.
+//
+// If no segment has available capacity, the pool waits a few times until memory becomes released
+// and falls back to allocating from the heap.
+type Pool struct {
+	maxSegments     int
+	segmentReleased chan struct{} // channel which is notified whenever any segment becomes available for allocations
+
+	// this protects the slice, to be able to atomically replace it
+	mu          sync.Mutex
+	segmentSize int
+	segments    []*segment
+}
+
+// NewPool creates a buffer pool, composed of N fixed-length segments of specified maximum size.
+func NewPool(segmentSize, maxSegments int) *Pool {
+	p := &Pool{
+		segmentSize:     segmentSize,
+		segmentReleased: make(chan struct{}),
+		maxSegments:     maxSegments,
+	}
+
+	return p
+}
+
+func (p *Pool) currentSegments() []*segment {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return p.segments
+}
+
+// SetSegmentSize sets the segment size for future segments that will be created.
+func (p *Pool) SetSegmentSize(maxSize int) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.segmentSize = maxSize
+}
+
+// InitializeSegments initializes n segments up to the maximum number of segments.
+func (p *Pool) InitializeSegments(n int) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	remaining := p.maxSegments - len(p.segments)
+	if n > remaining {
+		n = remaining
+	}
+
+	if n == 0 {
+		return false
+	}
+
+	var newSegments []*segment
+
+	newSegments = append(newSegments, p.segments...)
+
+	for i := 0; i < n; i++ {
+		newSegments = append(newSegments, &segment{
+			data: make([]byte, p.segmentSize),
+			pool: p,
+		})
+	}
+
+	p.segments = newSegments
+
+	return true
+}
+
+// Allocate allocates a slice of the buffer
+func (p *Pool) Allocate(n int) Buf {
+	// requested more than the pool can cache, allocate throw-away buffer.
+	if p == nil || n > p.segmentSize {
+		return Buf{0, 0, make([]byte, n), nil}
+	}
+
+	for i := 0; i < maxAllocAttempts; i++ {
+		// try to allocate
+		for _, s := range p.currentSegments() {
+			buf, ok := s.allocate(n)
+			if ok {
+				return buf
+			}
+		}
+
+		// add one more segment up to specified limit
+		if p.InitializeSegments(1) {
+			continue
+		}
+
+		// wait until some segment becomes free or some time passes, whichever comes first
+		select {
+		case <-p.segmentReleased:
```

Comment:
If this is a frequent code path and `<-p.segmentReleased` is a more frequent case than the one below, this ends up allocating many timers that are not released until the timer fires. Now, this may not make a difference given the `allocationRetryTimeout` and GC frequency.

Reply:
removed this whole channel business, we can simply add more segments